In [1]:
import pandas as pd
import numpy as np
import requests

# BeautifulSoup
from bs4 import BeautifulSoup
import lxml   # for parser

<div> 
    <h1 style="color: Red;"> Why did we import time? </h1>
    <p style = "margin-top: -15px; margin-left: 5px;;"> To make some delay to fetch data from JavaScript Dynamic web, if we go rush then website might think we are not human, we are bot who
    is trying to get everything from page and which can block our IP Adress from his website so that we will unable to go through the web
    again due to blockage. </p>
</div>

<div>
<h1 style="color: Red;"> What is Parser? </h1>
<p style = "margin-top: -15px; margin-left: 5px;"> Parser is a tool/software component that reads raw data (HTML, XML, JSON, etc.) and converts it into a structured format so that we can access that data easily. </p>
</div>

In [2]:
webpage = requests.get('https://www.ambitionbox.com/list-of-companies?page=1')

In [3]:
webpage

<Response [403]>

<h1 style="color:red;"> in above output - if response code is 403 </h1>
<h2 style="color: #41dc8e;"> Use This: </h2>
headers = {'User-Agent':'Mozilla/5.0 (Windows NT 6.3; Win 64 ; x64) Apple WeKit /537.36(KHTML , like Gecko) Chrome/80.0.3987.162 Safari/537.36'}

In [4]:
# Import requests library
import requests

# Define headers properly
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.162 Safari/537.36'
}

# Send request
webpage = requests.get(
    'https://www.ambitionbox.com/list-of-companies?page=1',
    headers=headers
).text

print(webpage[:500])   # First 500 characters check karne ke liye

<!doctype html>
<html data-n-head-ssr lang="en" data-n-head="%7B%22lang%22:%7B%22ssr%22:%22en%22%7D%7D">
  <head >
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width,initial-scale=1,minimum-scale=1">
    <meta http-equiv="X-UA-Compatible" content="IE=edge"> 
    <link rel="manifest" href="/assets/next/manifest.json">
    <style>@media only screen and (min-width:767px){.trp-img{width:400px!important;max-width:400px!important}}</style>
    <script src="/static/js/env-


In [5]:
soup = BeautifulSoup(webpage, 'lxml')

In [6]:
# it help to show HTML code in proper structred manner 

# code: 
#print(soup.prettify())

In [7]:
soup.find_all('h1')  # find all show all the value 

[<h1 class="companyListing__collectionTopHeadingCont" data-v-507d085f=""><div class="companyListing__collectionTopHeading" data-v-507d085f="">
 					Top Companies in
 				</div> <div class="companyListing__collectionIndiaHeading" data-v-507d085f=""><img data-v-507d085f="" height="12" src="https://static.ambitionbox.com/static/loc/loc-star.png" width="12"/>
 					INDIA
 					<img data-v-507d085f="" height="12" src="https://static.ambitionbox.com/static/loc/loc-star.png" width="12"/></div></h1>]

In [8]:
# to get the all text from <h1> tag
soup.find_all('h1')[0].get_text(separator=" ", strip=True)

'Top Companies in INDIA'

<div> 
    <h1 style = "color: #00CED1; margin-left: 5px;"> To Find Out The Name of Companies </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

In [9]:
for i in soup.find_all('h2')[1:]:  # used string slicing becuase heading of companies name where also a h2 tag and the heading is being gathering while we only want company names. to avoide heading of company names i used slicing which is basically skips the first value which is heading name, after that all the remaining names are companies name. 
    print(i.text.strip())

TCS
Accenture
Wipro
Cognizant
Capgemini
HDFC Bank
Infosys
ICICI Bank
HCLTech
Tech Mahindra
Genpact
Teleperformance
Jio
Axis Bank
Concentrix Corporation
Amazon
iEnergizer
Reliance Retail
LTIMindtree
IBM
Popular Collections by Industries
Popular Collections by Cities
Popular Collections by Roles


In [10]:
len(soup.find_all('meta', itemprop='name'))

20

In [11]:
len(soup.find_all('meta', attrs={'itemprop': 'name'}))   # attrs = attributes

20

<div> 
    <h1 style="color: #FFBF00;"> To Find Out The Ratings </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

In [12]:
for i in soup.find_all('div', attrs={"style": "height:auto;padding-bottom:1px;"}):
    print(i.text.strip())

3.3
3.7
3.6
3.6
3.7
3.8
3.5
4.0
3.4
3.4
3.6
3.8
4.4
3.6
3.6
3.9
4.6
3.9
3.6
3.9


<div> 
    <h1 style="color: #FFBF00;">  To Find Out The Number of Reviews </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

In [13]:
for i in soup.find_all('span', class_='companyCardWrapper__companyRatingCount'): 
    print(i.text.strip())

(1.1L)
(72.1k)
(64k)
(60.4k)
(52k)
(51.2k)
(47.8k)
(45.3k)
(45k)
(42.7k)
(41.3k)
(37.2k)
(33.3k)
(32.6k)
(31.6k)
(30.9k)
(27.2k)
(27.1k)
(26.1k)
(25.4k)


<div> 
    <h1 style="color: #FFBF00;">  To Find Out The Company Type </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

In [14]:
# in the output we are getting Location along with Company Type so next code is the solution to get the only industry type
for i in soup.find_all('span', class_='companyCardWrapper__interLinking'):
    print(i.text.strip())  

IT Services & Consulting | Bengaluru +439 other locations
IT Services & Consulting | Bengaluru +255 other locations
IT Services & Consulting | Hyderabad +371 other locations
IT Services & Consulting | Hyderabad +232 other locations
IT Services & Consulting | Bengaluru +182 other locations
Banking | Mumbai +1836 other locations
IT Services & Consulting | Bengaluru +248 other locations
Banking | Mumbai +1440 other locations
IT Services & Consulting | Chennai +231 other locations
IT Services & Consulting | Hyderabad +330 other locations
Analytics & KPO | Hyderabad +181 other locations
BPO | Mumbai +255 other locations
Telecom | Mumbai +1962 other locations
Banking | Mumbai +1509 other locations
BPO | Bengaluru +178 other locations
Internet | Bengaluru +520 other locations
BPO | Noida +52 other locations
Retail | Mumbai +1157 other locations
IT Services & Consulting | Bengaluru +145 other locations
IT Services & Consulting | Bengaluru +161 other locations


In [15]:
# Final answer code
head_quaeter = []
for i in soup.find_all('span', class_='companyCardWrapper__interLinking'):
    head_quaeter.append(i.text.split('|')[0].strip())

print(head_quaeter)

['IT Services & Consulting', 'IT Services & Consulting', 'IT Services & Consulting', 'IT Services & Consulting', 'IT Services & Consulting', 'Banking', 'IT Services & Consulting', 'Banking', 'IT Services & Consulting', 'IT Services & Consulting', 'Analytics & KPO', 'BPO', 'Telecom', 'Banking', 'BPO', 'Internet', 'BPO', 'Retail', 'IT Services & Consulting', 'IT Services & Consulting']


<div> 
    <h1 style="color: #FFBF00;">  To Find Out The Head Quarters </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

In [16]:
industries = []
for i in soup.find_all('span', class_='companyCardWrapper__interLinking'):
    industries.append(i.text.split('|')[1].strip())

print(industries)

['Bengaluru +439 other locations', 'Bengaluru +255 other locations', 'Hyderabad +371 other locations', 'Hyderabad +232 other locations', 'Bengaluru +182 other locations', 'Mumbai +1836 other locations', 'Bengaluru +248 other locations', 'Mumbai +1440 other locations', 'Chennai +231 other locations', 'Hyderabad +330 other locations', 'Hyderabad +181 other locations', 'Mumbai +255 other locations', 'Mumbai +1962 other locations', 'Mumbai +1509 other locations', 'Bengaluru +178 other locations', 'Bengaluru +520 other locations', 'Noida +52 other locations', 'Mumbai +1157 other locations', 'Bengaluru +145 other locations', 'Bengaluru +161 other locations']


<h2 style = "color: Red; margin-left: 10px;"> ⚠️The remaining processes cannot be done because website has updated and secured now which is not allowing us to get their data. (Uncompleted = 2 data col) </h2>

<div> 
    <h1 style="color: #FFBF00;">  To Find Out The Company Age </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

<div> 
    <h1 style="color: #FFBF00;">  To Find Out The Number of Employees </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

<h1 style="color: #BB86FC;"> Considering the Whole Container </h1>

In [17]:
company = soup.find_all('div', class_='companyCardWrapper')

In [18]:
len(company) # 24:40

20

In [28]:
name = []

for i in company:
    name.append(i.find('h2').text.strip())
    rating.append(i.find('h2'))



['TCS',
 'Accenture',
 'Wipro',
 'Cognizant',
 'Capgemini',
 'HDFC Bank',
 'Infosys',
 'ICICI Bank',
 'HCLTech',
 'Tech Mahindra',
 'Genpact',
 'Teleperformance',
 'Jio',
 'Axis Bank',
 'Concentrix Corporation',
 'Amazon',
 'iEnergizer',
 'Reliance Retail',
 'LTIMindtree',
 'IBM']

In [19]:
name = []
rating = []
reviews = []
ctype = []
hq = []
# old = []
# no_of_employee = []

for i in company:
    name.append(i.find('h2').text.strip())
    rating.append(i.find('div', attrs={"style": "height:auto;padding-bottom:1px;"}).text.strip())
    reviews.append(i.find('span', class_='companyCardWrapper__companyRatingCount').text.strip())
    ctype.append(i.find('span', class_='companyCardWrapper__interLinking').text.split('|')[0].strip())
    hq.append(i.find('span', class_='companyCardWrapper__interLinking').text.split('|')[1].strip())
    # old.append(i.find_all(ratings_list)[0][2].text.strip())
    # no_of_employee.append(i.find_all(ratings_list)[0][3].text.strip())



df=pd.DataFrame({'name':name,
   'rating':rating,
   'reviews':reviews,
   'company_type':ctype,
   'Head_Quarters':hq,
   # 'Company_Age':how_old,
   # 'No_of_Employee':no_of_employee,
   })

In [20]:
name

['TCS',
 'Accenture',
 'Wipro',
 'Cognizant',
 'Capgemini',
 'HDFC Bank',
 'Infosys',
 'ICICI Bank',
 'HCLTech',
 'Tech Mahindra',
 'Genpact',
 'Teleperformance',
 'Jio',
 'Axis Bank',
 'Concentrix Corporation',
 'Amazon',
 'iEnergizer',
 'Reliance Retail',
 'LTIMindtree',
 'IBM']

In [21]:
rating

['3.3',
 '3.7',
 '3.6',
 '3.6',
 '3.7',
 '3.8',
 '3.5',
 '4.0',
 '3.4',
 '3.4',
 '3.6',
 '3.8',
 '4.4',
 '3.6',
 '3.6',
 '3.9',
 '4.6',
 '3.9',
 '3.6',
 '3.9']

In [22]:
reviews

['(1.1L)',
 '(72.1k)',
 '(64k)',
 '(60.4k)',
 '(52k)',
 '(51.2k)',
 '(47.8k)',
 '(45.3k)',
 '(45k)',
 '(42.7k)',
 '(41.3k)',
 '(37.2k)',
 '(33.3k)',
 '(32.6k)',
 '(31.6k)',
 '(30.9k)',
 '(27.2k)',
 '(27.1k)',
 '(26.1k)',
 '(25.4k)']

In [23]:
df  # final scraped data

,name,rating,reviews,company_type,Head_Quarters
0,TCS,3.3,(1.1L),IT Services & Consulting,Bengaluru +439 other locations
1,Accenture,3.7,(72.1k),IT Services & Consulting,Bengaluru +255 other locations
2,Wipro,3.6,(64k),IT Services & Consulting,Hyderabad +371 other locations
3,Cognizant,3.6,(60.4k),IT Services & Consulting,Hyderabad +232 other locations
4,Capgemini,3.7,(52k),IT Services & Consulting,Bengaluru +182 other locations
5,HDFC Bank,3.8,(51.2k),Banking,Mumbai +1836 other locations
6,Infosys,3.5,(47.8k),IT Services & Consulting,Bengaluru +248 other locations
7,ICICI Bank,4.0,(45.3k),Banking,Mumbai +1440 other locations
8,HCLTech,3.4,(45k),IT Services & Consulting,Chennai +231 other locations
9,Tech Mahindra,3.4,(42.7k),IT Services & Consulting,Hyderabad +330 other locations


<div> 
    <h1 style="color: #FFD700;">  Creating DataFrame for all the Pages  </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

In [78]:
# 1. Create an empty list at the very outside (to add dataframes)
all_pages_data = []

for j in range(1, 1001): 

    url = f'https://www.ambitionbox.com/list-of-companies?page={j}'

    # Define headers properly
    headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.162 Safari/537.36'}
    
    # Send request
    webpage = requests.get(url, headers=headers).text
    soup = BeautifulSoup(webpage, 'lxml')
    company = soup.find_all('div', class_='companyCardWrapper')

    name, rating, reviews, ctype, hq = [], [], [], [], []
    
    
    for i in company:
        try:
            name.append(i.find('h2').text.strip())
        except: name.append(np.nan)

        try:
            # FIX 3: .split() + .join() logic for cleaner output
            r = i.find('div', attrs={"style": "height:auto;padding-bottom:1px;"}).text
            rating.append(" ".join(r.split()))
        except: rating.append(np.nan) # Spelling fix: appned -> append

        try:
            rev = i.find('span', class_='companyCardWrapper__companyRatingCount').text
            reviews.append(" ".join(rev.split()))
        except: reviews.append(np.nan)

        try:
            interlinking = i.find('span', class_='companyCardWrapper__interLinking').text.split('|')
            ctype.append(interlinking[0].strip())
            hq.append(interlinking[1].strip() if len(interlinking) > 1 else np.nan)
        except:
            ctype.append(np.nan)
            hq.append(np.nan)
    
    
    
    # Current page data dictionary
    d = {'name': name, 'rating': rating, 'reviews': reviews, 'company_type': ctype, 'Head_Quarters': hq}
    
    # Create dataframes inside the loop and add them to the list
    df = pd.DataFrame(d)
    all_pages_data.append(df)

# Add them all together after the loop ends (Anti-AttributeError)
final = pd.concat(all_pages_data, ignore_index=True)

In [81]:
# Final Result Check
final.shape

(10000, 5)

<div> 
    <h1 style="color: #FFD700;">  Creating and Saving all Fetched Data as .csv File  </h1>
    <p style = "margin-top: -15px; margin-left: 5px; opacity: 0.8;"> Web Scraped Data Output </p> 
</div>

In [82]:
# 1. Check for and remove duplicate rows (Clean data is preferred on Kaggle)
# We use the 'name' column as the unique identifier for companies
final_clean = final.drop_duplicates(subset=['name'])

# 2. Export the DataFrame to a CSV file
# index=False is essential to prevent an extra 'Unnamed: 0' column in our dataset
final_clean.to_csv('ambitionbox_companies_10k.csv', index=False)

# 3. Print final confirmation with the count of unique companies
print(f"Congratulations! The file is ready. Total unique companies: {len(final_clean)}")


Congratulations! The file is ready. Total unique companies: 9221
